packages/@aws-cdk/aws-codedeploy/lib/deployment-group.ts

```diff
@@ -141,19 +172,84 @@ export class ServerDeploymentGroup extends ServerDeploymentGroupRef {
         });
+        this._autoScalingGroups = (props && props.autoScalingGroups) || [];
+        this.addCodeDeployAgentInstallUserData(props && props.agentInstallationType);
 
         const resource = new cloudformation.DeploymentGroupResource(this, 'Resource', {
             applicationName: this.application.applicationName,
             deploymentGroupName: props && props.deploymentGroupName,
             serviceRoleArn: this.role.roleArn,
             deploymentConfigName: props && props.deploymentConfig &&
                 props.deploymentConfig.deploymentConfigName,
+            autoScalingGroups: props && props.autoScalingGroups &&
+                props.autoScalingGroups.map(asg => asg.autoScalingGroupName()),
         });
 
         this.deploymentGroupName = resource.ref;
         this.deploymentGroupArn = deploymentGroupName2Arn(this.application.applicationName,
             this.deploymentGroupName);
     }
+
+    public get autoScalingGroups(): autoscaling.AutoScalingGroup[] {
+        return this._autoScalingGroups.slice();
+    }
+
+    private addCodeDeployAgentInstallUserData(agentInstallationType?: AgentInstallationType): void {
+        if (!agentInstallationType) {
+            return;
+        }
+
+        const stack = cdk.Stack.find(this);
+        const region = stack.requireRegion('Could not determine the region to download the CodeDeploy agent from');
+
+        for (const asg of this._autoScalingGroups) {
+            switch (agentInstallationType) {
+                case AgentInstallationType.UBUNTU_14_04:
+                    asg.addUserData(`
+                        apt-get update -y
+                        apt-get install -y ruby2.0 wget
+                        TMP_DIR=\`mktemp -d\`
+                        cd $TMP_DIR
+                        wget https://aws-codedeploy-${region}.s3.amazonaws.com/latest/install
+                        chmod +x ./install
+                        ./install auto
+                        rm -fr $TMP_DIR
+                    `);
+                    break;
+                case AgentInstallationType.UBUNTU_16_04:
+                    asg.addUserData(`
+                        apt-get update -y
+                        apt-get install -y ruby wget
+                        TMP_DIR=\`mktemp -d\`
+                        cd $TMP_DIR
+                        wget https://aws-codedeploy-${region}.s3.amazonaws.com/latest/install
+                        chmod +x ./install
+                        ./install auto
+                        rm -fr $TMP_DIR
+                    `);
+                    break;
+                case AgentInstallationType.AMAZON_LINUX_RHEL:
+                    asg.addUserData(`
+                        yum update -y
+                        yum install ruby wget -y
+                        TMP_DIR=\`mktemp -d\`
+                        cd $TMP_DIR
+                        wget https://aws-codedeploy-${region}.s3.amazonaws.com/latest/install
```

Comment:
We should opt to move all of these to the `aws s3 cp` version - the url pattern above won't work across regions/partitions. The AWS CLI should  handle the right endpoint selection even in _fun_ regions.

Reply:
Will change to using the AWS CLI.